File: Python/Python/basics.py
"""
Python Basics Tutorial
A comprehensive guide to fundamental Python concepts
"""

# ============================================
# 1. VARIABLES AND DATA TYPES
# ============================================

# Numbers
integer_num = 42
float_num = 3.14
complex_num = 2 + 3j

print("=== Numbers ===")
print(f"Integer: {integer_num}, Type: {type(integer_num)}")
print(f"Float: {float_num}, Type: {type(float_num)}")
print(f"Complex: {complex_num}, Type: {type(complex_num)}")

# Strings
single_quote = 'Hello'
double_quote = "World"
multi_line = """This is a
multi-line string"""

print("\n=== Strings ===")
print(f"Single quotes: {single_quote}")
print(f"Double quotes: {double_quote}")
print(f"String concatenation: {single_quote + ' ' + double_quote}")
print(f"String repetition: {'Hi' * 3}")
print(f"String length: {len(single_quote)}")

# Boolean
is_true = True
is_false = False

print("\n=== Booleans ===")
print(f"True value: {is_true}")
print(f"False value: {is_false}")
print(f"Boolean from comparison: {10 > 5}")

# ============================================
# 2. COLLECTIONS
# ============================================

# Lists (mutable, ordered)
fruits = ["apple", "banana", "orange"]
mixed_list = [1, "hello", 3.14, True]

print("\n=== Lists ===")
print(f"Fruits list: {fruits}")
print(f"First fruit: {fruits[0]}")
print(f"Last fruit: {fruits[-1]}")
fruits.append("grape")
print(f"After append: {fruits}")
fruits.remove("banana")
print(f"After remove: {fruits}")

# Tuples (immutable, ordered)
coordinates = (10, 20)
rgb = (255, 128, 0)

print("\n=== Tuples ===")
print(f"Coordinates: {coordinates}")
print(f"X coordinate: {coordinates[0]}")
print(f"RGB values: {rgb}")

# Dictionaries (mutable, key-value pairs)
person = {
    "name": "Alice",
    "age": 30,
    "city": "New York"
}

print("\n=== Dictionaries ===")
print(f"Person dict: {person}")
print(f"Name: {person['name']}")
person["email"] = "alice@example.com"
print(f"After adding email: {person}")

# Sets (mutable, unique elements, unordered)
unique_nums = {1, 2, 3, 3, 4}
programming_langs = {"Python", "Java", "JavaScript"}

print("\n=== Sets ===")
print(f"Unique numbers: {unique_nums}")
print(f"Programming languages: {programming_langs}")
programming_langs.add("C++")
print(f"After adding C++: {programming_langs}")

# ============================================
# 3. OPERATORS
# ============================================

print("\n=== Operators ===")

# Arithmetic operators
a, b = 10, 3
print(f"Addition: {a} + {b} = {a + b}")
print(f"Subtraction: {a} - {b} = {a - b}")
print(f"Multiplication: {a} * {b} = {a * b}")
print(f"Division: {a} / {b} = {a / b}")
print(f"Floor division: {a} // {b} = {a // b}")
print(f"Modulus: {a} % {b} = {a % b}")
print(f"Power: {a} ** {b} = {a ** b}")

# Comparison operators
x, y = 5, 8
print(f"\nComparison: x={x}, y={y}")
print(f"x == y: {x == y}")
print(f"x != y: {x != y}")
print(f"x < y: {x < y}")
print(f"x <= y: {x <= y}")

# Logical operators
p, q = True, False
print(f"\nLogical: p={p}, q={q}")
print(f"p and q: {p and q}")
print(f"p or q: {p or q}")
print(f"not p: {not p}")

# ============================================
# 4. CONTROL FLOW
# ============================================

print("\n=== Control Flow ===")

# If-elif-else
score = 85
print(f"Score: {score}")
if score >= 90:
    grade = "A"
elif score >= 80:
    grade = "B"
elif score >= 70:
    grade = "C"
else:
    grade = "F"
print(f"Grade: {grade}")

# For loops
print("\nFor loop - iterating through list:")
colors = ["red", "green", "blue"]
for color in colors:
    print(f"  Color: {color}")

print("\nFor loop with range:")
for i in range(3):
    print(f"  Count: {i}")

print("\nFor loop with enumerate:")
for index, color in enumerate(colors):
    print(f"  {index}: {color}")

# While loop
print("\nWhile loop:")
counter = 0
while counter < 3:
    print(f"  Counter: {counter}")
    counter += 1

# List comprehension
numbers = [1, 2, 3, 4, 5]
squares = [n ** 2 for n in numbers]
print(f"\nList comprehension:")
print(f"Original: {numbers}")
print(f"Squares: {squares}")

even_numbers = [n for n in numbers if n % 2 == 0]
print(f"Even numbers: {even_numbers}")

# ============================================
# 5. FUNCTIONS
# ============================================

print("\n=== Functions ===")

# Basic function
def greet(name):
    """Simple greeting function"""
    return f"Hello, {name}!"

print(greet("Bob"))

# Function with default parameter
def power(base, exponent=2):
    """Calculate base raised to exponent (default: 2)"""
    return base ** exponent

print(f"power(3): {power(3)}")
print(f"power(3, 3): {power(3, 3)}")

# Function with multiple returns
def calculate_stats(numbers):
    """Calculate sum and average of numbers"""
    total = sum(numbers)
    avg = total / len(numbers) if numbers else 0
    return total, avg

nums = [10, 20, 30, 40, 50]
total, average = calculate_stats(nums)
print(f"Numbers: {nums}")
print(f"Total: {total}, Average: {average}")

# Lambda function
square = lambda x: x ** 2
print(f"\nLambda function - square(5): {square(5)}")

# Map, filter, reduce
from functools import reduce

numbers = [1, 2, 3, 4, 5]
squared = list(map(lambda x: x ** 2, numbers))
filtered = list(filter(lambda x: x > 2, numbers))
sum_all = reduce(lambda x, y: x + y, numbers)

print(f"Original: {numbers}")
print(f"Mapped (squared): {squared}")
print(f"Filtered (>2): {filtered}")
print(f"Reduced (sum): {sum_all}")

# ============================================
# 6. CLASSES AND OBJECTS
# ============================================

print("\n=== Classes and Objects ===")

class Dog:
    """Simple Dog class"""
    
    # Class variable
    species = "Canis familiaris"
    
    def __init__(self, name, age):
        """Constructor method"""
        self.name = name  # Instance variable
        self.age = age
    
    def bark(self):
        """Method to make dog bark"""
        return f"{self.name} says Woof!"
    
    def describe(self):
        """Method to describe the dog"""
        return f"{self.name} is {self.age} years old"

# Create instances
dog1 = Dog("Buddy", 3)
dog2 = Dog("Max", 5)

print(f"Dog 1: {dog1.describe()}")
print(f"Dog 1 barks: {dog1.bark()}")
print(f"Dog 2: {dog2.describe()}")
print(f"Species: {Dog.species}")

# Inheritance
class Puppy(Dog):
    """Puppy class inheriting from Dog"""
    
    def __init__(self, name, age, training_level):
        super().__init__(name, age)
        self.training_level = training_level
    
    def play(self):
        return f"{self.name} is playing!"

puppy = Puppy("Charlie", 1, "beginner")
print(f"\nPuppy: {puppy.describe()}")
print(f"Puppy plays: {puppy.play()}")

# ============================================
# 7. EXCEPTION HANDLING
# ============================================

print("\n=== Exception Handling ===")

# Basic try-except
try:
    result = 10 / 2
    print(f"10 / 2 = {result}")
except ZeroDivisionError:
    print("Cannot divide by zero!")

# Multiple exceptions and else
def safe_divide(a, b):
    try:
        result = a / b
    except ZeroDivisionError:
        return "Error: Division by zero"
    except TypeError:
        return "Error: Invalid types for division"
    else:
        return f"Result: {result}"
    finally:
        print("Division operation completed")

print(safe_divide(10, 2))
print(safe_divide(10, 0))

# ============================================
# 8. FILE OPERATIONS
# ============================================

print("\n=== File Operations (Examples) ===")

# Writing to file (example - not executed)
example_write = """
with open('example.txt', 'w') as file:
    file.write('Hello, World!')
    file.write('\\nPython is awesome!')
"""
print("Writing to file:")
print(example_write)

# Reading from file (example - not executed)
example_read = """
with open('example.txt', 'r') as file:
    content = file.read()
    print(content)
"""
print("\nReading from file:")
print(example_read)

# ============================================
# 9. MODULES AND IMPORTS
# ============================================

print("\n=== Modules and Imports ===")

import math
import random
from datetime import datetime

print(f"Math.pi: {math.pi}")
print(f"Square root of 16: {math.sqrt(16)}")
print(f"Random number (1-10): {random.randint(1, 10)}")
print(f"Current time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

# ============================================
# 10. USEFUL BUILT-IN FUNCTIONS
# ============================================

print("\n=== Useful Built-in Functions ===")

numbers = [3, 1, 4, 1, 5, 9, 2, 6]
print(f"Numbers: {numbers}")
print(f"Length: {len(numbers)}")
print(f"Sum: {sum(numbers)}")
print(f"Max: {max(numbers)}")
print(f"Min: {min(numbers)}")
print(f"Sorted: {sorted(numbers)}")
print(f"Reversed: {list(reversed(numbers))}")

# Type conversion
num_str = "123"
print(f"\nType conversion:")
print(f"String '{num_str}' to int: {int(num_str)}")
print(f"Int 456 to string: {str(456)}")
print(f"String '3.14' to float: {float('3.14')}")

# Zip function
names = ["Alice", "Bob", "Charlie"]
ages = [25, 30, 35]
combined = list(zip(names, ages))
print(f"\nZip function:")
print(f"Names: {names}")
print(f"Ages: {ages}")
print(f"Zipped: {combined}")

# Any and all
bool_list = [True, True, False]
print(f"\nBoolean operations on {bool_list}:")
print(f"Any true? {any(bool_list)}")
print(f"All true? {all(bool_list)}")

print("\n" + "="*50)
print("Python Basics Tutorial Complete!")
print("="*50)

File: CPP/basics.cpp
// ============================================
// BASIC C++ TUTORIAL WITH EXAMPLES
// ============================================

#include <iostream>  // For input/output operations
#include <string>    // For string operations
#include <vector>    // For dynamic arrays

using namespace std;  // Use standard namespace

// ============================================
// 1. BASIC STRUCTURE & HELLO WORLD
// ============================================

void helloWorld() {
    // cout is used for output, endl adds a new line
    cout << "Hello, World!" << endl;
}

// ============================================
// 2. VARIABLES & DATA TYPES
// ============================================

void variablesAndTypes() {
    // Integer types
    int age = 25;           // Whole numbers
    long population = 1000000L;
    
    // Floating point types
    float price = 19.99f;   // Single precision
    double pi = 3.14159;    // Double precision (more accurate)
    
    // Character and boolean
    char grade = 'A';       // Single character
    bool isPassed = true;   // true or false
    
    // String (requires #include <string>)
    string name = "John";
    
    // Display values
    cout << "Name: " << name << ", Age: " << age << endl;
    cout << "Price: $" << price << ", Grade: " << grade << endl;
}

// ============================================
// 3. OPERATORS
// ============================================

void operators() {
    int a = 10, b = 3;
    
    // Arithmetic operators
    cout << "Addition: " << (a + b) << endl;        // 13
    cout << "Subtraction: " << (a - b) << endl;     // 7
    cout << "Multiplication: " << (a * b) << endl;  // 30
    cout << "Division: " << (a / b) << endl;        // 3 (integer division)
    cout << "Modulus: " << (a % b) << endl;         // 1 (remainder)
    
    // Comparison operators (return bool)
    cout << "Is a > b? " << (a > b) << endl;   // 1 (true)
    cout << "Is a == b? " << (a == b) << endl; // 0 (false)
    
    // Logical operators
    bool x = true, y = false;
    cout << "x AND y: " << (x && y) << endl;  // 0 (false)
    cout << "x OR y: " << (x || y) << endl;   // 1 (true)
    cout << "NOT x: " << (!x) << endl;        // 0 (false)
}

// ============================================
// 4. CONDITIONAL STATEMENTS
// ============================================

void conditionals() {
    int score = 85;
    
    // if-else statement
    if (score >= 90) {
        cout << "Grade: A" << endl;
    } else if (score >= 80) {
        cout << "Grade: B" << endl;
    } else if (score >= 70) {
        cout << "Grade: C" << endl;
    } else {
        cout << "Grade: F" << endl;
    }
    
    // Switch statement
    int day = 3;
    switch (day) {
        case 1:
            cout << "Monday" << endl;
            break;  // Break prevents fall-through
        case 2:
            cout << "Tuesday" << endl;
            break;
        case 3:
            cout << "Wednesday" << endl;
            break;
        default:
            cout << "Other day" << endl;
    }
}

// ============================================
// 5. LOOPS
// ============================================

void loops() {
    // For loop - used when you know iteration count
    cout << "For loop: ";
    for (int i = 1; i <= 5; i++) {
        cout << i << " ";  // Prints: 1 2 3 4 5
    }
    cout << endl;
    
    // While loop - used when condition-based
    cout << "While loop: ";
    int j = 1;
    while (j <= 5) {
        cout << j << " ";
        j++;
    }
    cout << endl;
    
    // Do-while loop - executes at least once
    cout << "Do-while loop: ";
    int k = 1;
    do {
        cout << k << " ";
        k++;
    } while (k <= 5);
    cout << endl;
}

// ============================================
// 6. ARRAYS
// ============================================

void arrays() {
    // Fixed-size array
    int numbers[5] = {10, 20, 30, 40, 50};
    
    cout << "Array elements: ";
    for (int i = 0; i < 5; i++) {
        cout << numbers[i] << " ";
    }
    cout << endl;
    
    // Multidimensional array
    int matrix[2][3] = {
        {1, 2, 3},
        {4, 5, 6}
    };
    
    cout << "Matrix:" << endl;
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 3; j++) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
}

// ============================================
// 7. FUNCTIONS
// ============================================

// Function that returns a value
int add(int x, int y) {
    return x + y;
}

// Function with no return value (void)
void greet(string name) {
    cout << "Hello, " << name << "!" << endl;
}

// Function with default parameter
int multiply(int a, int b = 2) {
    return a * b;
}

void functionsDemo() {
    int sum = add(5, 3);
    cout << "Sum: " << sum << endl;
    
    greet("Alice");
    
    cout << "Multiply: " << multiply(5) << endl;      // Uses default b=2
    cout << "Multiply: " << multiply(5, 4) << endl;   // Overrides default
}

// ============================================
// 8. PASS BY VALUE VS PASS BY REFERENCE
// ============================================

// Pass by value - creates a copy
void incrementValue(int x) {
    x++;  // Only changes the copy
}

// Pass by reference - works with original
void incrementReference(int &x) {
    x++;  // Changes the original variable
}

void passingDemo() {
    int num = 10;
    
    incrementValue(num);
    cout << "After pass by value: " << num << endl;  // Still 10
    
    incrementReference(num);
    cout << "After pass by reference: " << num << endl;  // Now 11
}

// ============================================
// 9. POINTERS
// ============================================

void pointers() {
    int value = 42;
    int *ptr = &value;  // ptr stores the address of value
    
    cout << "Value: " << value << endl;
    cout << "Address: " << &value << endl;
    cout << "Pointer: " << ptr << endl;
    cout << "Dereferenced pointer: " << *ptr << endl;  // Gets value at address
    
    // Modify through pointer
    *ptr = 100;
    cout << "New value: " << value << endl;  // 100
}

// ============================================
// 10. VECTORS (DYNAMIC ARRAYS)
// ============================================

void vectors() {
    // Vector - dynamic array that can grow/shrink
    vector<int> nums;
    
    // Add elements
    nums.push_back(10);
    nums.push_back(20);
    nums.push_back(30);
    
    cout << "Vector elements: ";
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i] << " ";
    }
    cout << endl;
    
    // Range-based for loop (C++11)
    cout << "Using range-based loop: ";
    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;
    
    // Remove last element
    nums.pop_back();
    cout << "After pop_back, size: " << nums.size() << endl;
}

// ============================================
// 11. CLASSES & OBJECTS (OOP)
// ============================================

class Rectangle {
private:
    // Private members - accessible only within class
    double width;
    double height;
    
public:
    // Constructor - called when object is created
    Rectangle(double w, double h) {
        width = w;
        height = h;
    }
    
    // Member functions (methods)
    double area() {
        return width * height;
    }
    
    double perimeter() {
        return 2 * (width + height);
    }
    
    // Getter methods
    double getWidth() { return width; }
    double getHeight() { return height; }
    
    // Setter methods
    void setWidth(double w) { width = w; }
    void setHeight(double h) { height = h; }
};

void classesDemo() {
    // Create object
    Rectangle rect(5.0, 3.0);
    
    cout << "Rectangle area: " << rect.area() << endl;
    cout << "Rectangle perimeter: " << rect.perimeter() << endl;
    
    // Modify using setters
    rect.setWidth(10.0);
    cout << "New area: " << rect.area() << endl;
}

// ============================================
// MAIN FUNCTION - PROGRAM ENTRY POINT
// ============================================

int main() {
    cout << "=== C++ TUTORIAL ===" << endl << endl;
    
    cout << "1. Hello World:" << endl;
    helloWorld();
    cout << endl;
    
    cout << "2. Variables & Types:" << endl;
    variablesAndTypes();
    cout << endl;
    
    cout << "3. Operators:" << endl;
    operators();
    cout << endl;
    
    cout << "4. Conditionals:" << endl;
    conditionals();
    cout << endl;
    
    cout << "5. Loops:" << endl;
    loops();
    cout << endl;
    
    cout << "6. Arrays:" << endl;
    arrays();
    cout << endl;
    
    cout << "7. Functions:" << endl;
    functionsDemo();
    cout << endl;
    
    cout << "8. Pass by Value vs Reference:" << endl;
    passingDemo();
    cout << endl;
    
    cout << "9. Pointers:" << endl;
    pointers();
    cout << endl;
    
    cout << "10. Vectors:" << endl;
    vectors();
    cout << endl;
    
    cout << "11. Classes & Objects:" << endl;
    classesDemo();
    
    return 0;  // Return 0 indicates successful execution
}